File: internal/ai/gemini.go
package ai

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
)

func GeneratePlaylistGemini(ctx context.Context, apiKey, prompt string, count int) ([]Song, error) {
	fullPrompt := fmt.Sprintf(`You are a music expert. Generate a playlist of exactly %d songs for: "%s"

Return ONLY a JSON array of songs, no other text. Each song should have "title" and "artist" fields.
Example: [{"title": "Blue in Green", "artist": "Miles Davis"}, {"title": "Take Five", "artist": "Dave Brubeck"}]

Return only the JSON array, no explanation or markdown formatting.`, count, prompt)

	payload := map[string]any{
		"contents": []map[string]any{
			{"parts": []map[string]string{{"text": fullPrompt}}},
		},
		"generationConfig": map[string]int{"maxOutputTokens": 2048},
	}
	buf, _ := json.Marshal(payload)
	u := "https://generativelanguage.googleapis.com/v1beta/models/gemini-3-flash-preview:generateContent?key=" + url.QueryEscape(apiKey)
	req, err := http.NewRequestWithContext(ctx, http.MethodPost, u, bytes.NewReader(buf))
	if err != nil {
		return nil, err
	}
	req.Header.Set("Content-Type", "application/json")

	resp, err := apiHTTPClient.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	body, _ := io.ReadAll(resp.Body)
	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		return nil, fmt.Errorf("gemini api error: %d - %s", resp.StatusCode, string(body))
	}

	var data struct {
		Candidates []struct {
			Content struct {
				Parts []struct {
					Text string `json:"text"`
				} `json:"parts"`
			} `json:"content"`
		} `json:"candidates"`
	}
	if err := json.Unmarshal(body, &data); err != nil {
		return nil, err
	}
	text := ""
	if len(data.Candidates) > 0 && len(data.Candidates[0].Content.Parts) > 0 {
		text = data.Candidates[0].Content.Parts[0].Text
	}
	return parsePlaylistResponse(text)
}

File: src/ai/gemini.ts
import { Song } from './index.js';

export async function generatePlaylistGemini(
  apiKey: string,
  prompt: string,
  count: number
): Promise<Song[]> {
  const fullPrompt = `You are a music expert. Generate a playlist of exactly ${count} songs for: "${prompt}"

Return ONLY a JSON array of songs, no other text. Each song should have "title" and "artist" fields.
Example: [{"title": "Blue in Green", "artist": "Miles Davis"}, {"title": "Take Five", "artist": "Dave Brubeck"}]

Return only the JSON array, no explanation or markdown formatting.`;

  const response = await fetch(
    `https://generativelanguage.googleapis.com/v1beta/models/gemini-3-flash-preview:generateContent?key=${apiKey}`,
    {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({
        contents: [{ parts: [{ text: fullPrompt }] }],
        generationConfig: {
          maxOutputTokens: 2048,
        },
      }),
    }
  );

  if (!response.ok) {
    const error = await response.text();
    throw new Error(`Gemini API error: ${response.status} - ${error}`);
  }

  const data = (await response.json()) as {
    candidates: Array<{ content: { parts: Array<{ text: string }> } }>;
  };
  const text = data.candidates[0]?.content?.parts[0]?.text || '';

  return parsePlaylistResponse(text);
}

function parsePlaylistResponse(text: string): Song[] {
  // Try direct JSON parse first
  try {
    const parsed = JSON.parse(text);
    if (Array.isArray(parsed)) {
      return parsed.filter(
        (s): s is Song =>
          typeof s === 'object' &&
          s !== null &&
          typeof s.title === 'string' &&
          typeof s.artist === 'string'
      );
    }
  } catch {
    // Fall through to regex extraction
  }

  // Try to extract JSON array from text
  const jsonMatch = text.match(/\[[\s\S]*\]/);
  if (jsonMatch) {
    try {
      const parsed = JSON.parse(jsonMatch[0]);
      if (Array.isArray(parsed)) {
        return parsed.filter(
          (s): s is Song =>
            typeof s === 'object' &&
            s !== null &&
            typeof s.title === 'string' &&
            typeof s.artist === 'string'
        );
      }
    } catch {
      // Continue to fallback
    }
  }

  throw new Error('Failed to parse playlist from AI response');
}
